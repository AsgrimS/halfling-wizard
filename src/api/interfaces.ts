export interface Source {
  index: string;
  name: string;
  url: string;
}
export interface ProficiencyChoices {
  choose: number;
  type: string;
  from: Source[];
}

export interface Hero {
  index: string;
  name: string;
  url: string;
}
