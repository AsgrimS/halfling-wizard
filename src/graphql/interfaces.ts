import {
  Class,
  Subclass,
  Proficiency,
  ClassProficiency_Choices,
  AbilityScoreSkills,
  ClassSpellcasting,
  ClassStarting_Equipment,
} from "./schema";

export interface HeroClassesData {
  classes: Class[];
}

export interface HeroClassData {
  class: Class;
}

export interface SubclassesData {
  subclasses: Subclass[];
}

export interface ProficienciesData {
  proficiencies: Proficiency[];
}

export interface ProficienciesChoicesData {
  proficienciesChoices: ClassProficiency_Choices[];
}
export interface SavingThrowsData {
  saving_throws: AbilityScoreSkills[];
}

export interface ClassSpellcastingData {
  spellcasting: ClassSpellcasting;
}

export interface StartingEquipmentData {
  startingEquipments: ClassStarting_Equipment[];
}
