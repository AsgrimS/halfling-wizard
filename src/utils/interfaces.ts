import { QueryLazyOptions, OperationVariables } from "@apollo/client";

export interface LazyQueryCallbackFunction {
  getClassDetail: (
    options?: QueryLazyOptions<OperationVariables> | undefined
  ) => void;
}
