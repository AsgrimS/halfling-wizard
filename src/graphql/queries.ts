import { gql } from "@apollo/client";

export const HERO_CLASS = gql`
  query GetHeroesClasses($classIndex: String!) {
    class(filter: { name: $classIndex }) {
      hit_die
      index
      name
      proficiencies {
        index
        name
      }
      saving_throws {
        index
        name
      }
      proficiency_choices {
        choose
        from {
          name
          index
        }
      }
      spellcasting {
        level
        spellcasting_ability {
          index
          name
        }
      }
      starting_equipment {
        equipment {
          name
          index
        }
        quantity
      }
      starting_equipment_options {
        choose
        from {
          equipment {
            name
            index
          }
          quantity
        }
      }
      subclasses {
        index
        name
      }
    }
  }
`;

export const HEROES_CLASSES = gql`
  query GetHeroesClasses {
    classes {
      name
      index
    }
  }
`;
