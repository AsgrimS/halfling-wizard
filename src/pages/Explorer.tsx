import React from "react";

import { useLazyQuery } from "@apollo/client";

import { HERO_CLASS } from "../graphql/queries";
import { HeroClassData } from "../graphql/interfaces";
import ClassesList from "../components/ClassesList";
import ClassDetails from "../components/ClassDetails";
import LoadingAnimation from "../components/styled-components/LoadingAnimation";

const Explorer: React.FC = () => {
  const [getClassDetail, heroClassData] = useLazyQuery<HeroClassData>(
    HERO_CLASS
  );
  return (
    <>
      <ClassesList getClassDetail={getClassDetail} />
      {heroClassData.called && heroClassData.loading === true && (
        <LoadingAnimation pxSize={24} />
      )}
      {heroClassData.loading === false && heroClassData.data && (
        <ClassDetails class={heroClassData.data.class} />
      )}
    </>
  );
};

export default Explorer;
