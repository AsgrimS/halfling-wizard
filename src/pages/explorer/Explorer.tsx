import React from "react";

import { useLazyQuery } from "@apollo/client";

import { HERO_CLASS } from "../../graphql/queries";
import { HeroClassData } from "../../graphql/interfaces";
import ClassesList from "../../components/classesList/ClassesList";
import ClassDetails from "../../components/classDetails/ClassDetails";
import LoadingAnimation from "../../components/styled-components/LoadingAnimation";

const Explorer: React.FC = () => {
  const [getClassDetail, heroClassData] = useLazyQuery<HeroClassData>(
    HERO_CLASS
  );

  return (
    <section data-testid="explorer">
      <ClassesList getClassDetail={getClassDetail} />
      {heroClassData.called && heroClassData.loading === true && (
        <LoadingAnimation pxSize={24} />
      )}
      {heroClassData.loading === false && heroClassData.data && (
        <ClassDetails class={heroClassData.data.class} />
      )}
    </section>
  );
};

export default Explorer;
