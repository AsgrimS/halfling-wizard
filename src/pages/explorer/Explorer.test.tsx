import React from "react";
import { act } from "react-dom/test-utils";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

import { Class } from "../../graphql/schema";
import { HERO_CLASS } from "../../graphql/queries";
import { LazyQueryCallbackFunction } from "../../utils/interfaces";

import Explorer from "./Explorer";

jest.mock("../../components/classDetails/ClassDetails", () => () => (
  <div data-testid="mock-class-details" />
));
jest.mock(
  "../../components/classesList/ClassesList",
  () => ({ getClassDetail }: LazyQueryCallbackFunction) => (
    <div data-testid="mock-classes-list">
      <button
        data-testid="mock-btn"
        type="button"
        onClick={() => getClassDetail()}
      >
        test
      </button>
    </div>
  )
);

describe("Explorer Page", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render a page that renders classes list", () => {
    render(
      <MockedProvider>
        <Explorer />
      </MockedProvider>
    );

    const explorerElement = screen.getByTestId("explorer");
    expect(explorerElement).toBeInTheDocument();

    const classesListMockElement = screen.getByTestId("mock-classes-list");
    expect(classesListMockElement).toBeInTheDocument();

    expect(explorerElement).toMatchSnapshot();
  });

  it("should render a page that renders classes list and class details components", async () => {
    const mockClassData: Class = {
      hit_die: 1,
      index: "wizard",
      name: "Wizard",
      proficiencies: [
        {
          index: "foo",
          name: "Foo",
        },
      ],
      saving_throws: [{ index: "int", name: "INT" }],
      proficiency_choices: [
        { choose: 1, from: [{ name: "Foo", index: "foo" }] },
      ],
      spellcasting: {
        level: 1,
        spellcasting_ability: {
          index: "foo",
          name: "Foo",
        },
      },
      starting_equipment: [
        { equipment: { name: "bar", index: "foo" }, quantity: 1 },
      ],
      starting_equipment_options: [
        {
          choose: 1,
          from: [{ equipment: { name: "bar", index: "foo" }, quantity: 1 }],
        },
      ],
      subclasses: [{ index: "foo", name: "bar" }],
    };
    const mockQuery: MockedResponse = {
      request: { query: HERO_CLASS },
      result: { data: { class: mockClassData } },
    };

    act(() => {
      render(
        <MockedProvider mocks={[mockQuery]} addTypename={false}>
          <Explorer />
        </MockedProvider>
      );
    });
    const explorerElement = screen.getByTestId("explorer");
    expect(explorerElement).toBeInTheDocument();

    const getQueryButton = screen.getByTestId("mock-btn");
    fireEvent.click(getQueryButton);
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    const classDetailsMockElement = screen.getByTestId("mock-class-details");
    expect(classDetailsMockElement).toBeInTheDocument();

    expect(explorerElement).toMatchSnapshot();
  });
});
