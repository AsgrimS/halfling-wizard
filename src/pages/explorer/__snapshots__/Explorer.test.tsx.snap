// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explorer Page should render a page that renders classes list 1`] = `
<section
  data-testid="explorer"
>
  <div
    data-testid="mock-classes-list"
  >
    <button
      data-testid="mock-btn"
      type="button"
    >
      test
    </button>
  </div>
</section>
`;

exports[`Explorer Page should render a page that renders classes list and class details components 1`] = `
<section
  data-testid="explorer"
>
  <div
    data-testid="mock-classes-list"
  >
    <button
      data-testid="mock-btn"
      type="button"
    >
      test
    </button>
  </div>
  <div
    data-testid="mock-class-details"
  />
</section>
`;
